qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,199 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+r"""
+=========
+Estimator
+=========
+
+Estimator class estimates expectation values of quantum circuits and observables.
+
+An estimator object is initialized with multiple quantum circuits and observables
+and users can specify pairs of quantum circuits and observables
+to estimate the expectation values.
+
+The input consists of following elements.
+
+* quantum circuits (:math:`\psi_i(\theta)`): list of (parametrized) quantum circuits
+  or state vectors to be converted into quantum circuits.
+  (a list of :class:`~qiskit.circuit.QuantumCircuit`))
+
+* observables (:math:`H_j`): a list of :class:`~qiskit.quantum_info.SparsePauliOp`.
+
+* grouping: a list of pairs :class:`~qiskit.primitive.Group` of an
```

Comment:
Here, it may not be user-friendly to pass the index as an integer.
For example, it would be easier to use a list of tuples of QuantumCircuit and SparsePauliOp. In the case of Runtime, I'm thinking of matching by object id and sending the index, rather than actually sending the objects to the server. I think QuantumCircuit and SparsePauliOperator would be fine for a simple interface.

Reply:
I see the point,  but it can be tricky because QuantumCircuit is mutable and not hashable, so there could be confusion if the circuit changes between uses in the initializer and run methods. If we would have an immutable circuit object I would support this

Reply:
I don't think we have to wait for immutable circuits to enable this (there is no near-term prospect for that). 

Yes, in the unlikely event that a circuit changes between calls, then this would raise an error. But for the majority of users who don't change the circuit, this just improves code readability by a lot. i.e. the syntax becomes much closer to the comment in this code:

```py
# calculate [ <psi1(theta1)|H2|psi1(theta1)>, <psi1(theta1)|H3|psi1(theta1)> ]
result = estimator([0, 0], [1, 2], [theta1] * 2)
```

Qiskit has a similar thing for `result.get_counts()` where you can pass the circuit index of your job or the circuit itself. I haven't seen people complain about the pitfall of changing circuits here, and I think passing circuits is the common usage for this function now.

Reply:
💯 to being able to pass objects instead of indices as an alternative that we should accept

Reply:
OK. I try to implement it after #7780 is merged.